### PR TITLE
Issue #495: Add a useful dashboard

### PR DESCRIPTION
**Pull request by docwilmot**
Fixes https://github.com/backdrop/backdrop-issues/issues/495
Replaces https://github.com/backdrop/backdrop/pull/2400

### CODE REVIEW

File: core/modules/dashboard/dashboard.module [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+function dashboard_menu() {
+  $items = array();
+  $dashboard = layout_load('dashboard');
+  if ($dashboard && !$dashboard->disabled) {
+    $items['admin/dashboard/settings'] = array(
+      'title' => 'Settings',
+      'page callback' => 'backdrop_get_form',
+      'page arguments' => array('dashboard_admin_settings'),
+      'description' => 'Admin dashboard settings',
+      'access arguments' => array('access content'),
```
Comment (quicksketch) [inline core/modules/dashboard/dashboard.module]: Seems this should have a different permission? Do we want/need "Administer dashboard" permission?

File: core/modules/dashboard/dashboard.module [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+function dashboard_user_login(&$edit, $account) {
+  // Don't redirect for password reset.
+  if (!((arg(0) == 'user') && (arg(1) == 'reset'))) {
+    // Check for permission to use dashboard.
+    if (user_access('access dashboard', $account)) {
+      $dashboard = layout_load('dashboard');
+      if ($dashboard && !$dashboard->disabled && $dashboard->checkAccess()) {
+        // Check for redirect to dashboard setting.
+        if (config_get('dashboard.settings', 'dashboard_login_redirect') == 1) {
+          $_GET['destination'] = 'admin/dashboard';
```
Comment (quicksketch) [inline core/modules/dashboard/dashboard.module]: Is this the same way that Login Redirect modules work? Seems a bit overpowered versus using a form_alter(). We should also at least check that we're not overriding an existing `destination`. e.g. if filling out a form at `user/login?destination=node/add` you should still be taken to the `node/add`, not the dashboard.

File: core/modules/dashboard/config/layout.layout.dashboard.json [hunk 3]
```diff
@@ -0,0 +1,353 @@
+{
+    "_config_name": "layout.layout.dashboard",
+    "path": "admin/dashboard",
+    "name": "dashboard",
+    "title": "Dashboard",
+    "description": "Site overview dashboard",
+    "renderer_name": "standard",
+    "module": "dashboard",
+    "weight": 0,
+    "storage": 2,
```
Comment (quicksketch) [inline core/modules/dashboard/config/layout.layout.dashboard.json]: Should be set to `4` for the default state.

File: core/modules/dashboard/config/layout.layout.dashboard.json [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        ],
+        "bottom": [],
+        "footer": [
+            "55aa5247-2948-450b-b65c-0fcff326e1a1"
+        ],
+        "title": []
+    },
+    "content": {
+        "1d07f2f1-c523-4c0d-a8ca-e7089aed712b": {
+            "plugin": "system:header",
```
Comment (quicksketch) [inline core/modules/dashboard/config/layout.layout.dashboard.json]: I think the system header block should be removed, to match the default administrative layout.

**Review by quicksketch — COMMENTED** [review]
Looking good so far! I like the blocks as being part of the module. Thanks for building out tests as we go.

File: core/modules/dashboard/dashboard.module (hunk 1 above)
Comment (opi) [reply core/modules/dashboard/dashboard.module]: +1 for "administer dashboard" permission.

File: core/modules/dashboard/dashboard.module (hunk 2 above)
Comment (docwilmot) [reply core/modules/dashboard/dashboard.module]: I've had a few looks at the other redirect modules and the internet. Setting `$_GET['destination']` seems a common practice. [Login destination module](https://cgit.drupalcode.org/login_destination/tree/login_destination.module#n158) mentions they use a form alter to allow external URLs, but we don't need that. 
I agree we should account for existing `$_GET['destination']s`.

File: core/modules/dashboard/dashboard.module [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+ */
+function dashboard_user_delete($account) {
+  cache_clear_all('dashboard_user_overview', 'cache');
+}
+
+/**
+ * Implements hook_user_login().
+ * Redirects to the dashboard if configured.
+ */
+function dashboard_user_login(&$edit, $account) {
```
Comment (herbdool) [inline core/modules/dashboard/dashboard.module:147]: There's a lot of nesting conditions going on here which makes it hard to understand. I don't know if Backdrop has a code style for this, but I've found this link helpful in making my conditionals easier to read: https://carlalexander.ca/mastering-php-conditionals/

Here's an alternative version (I haven't tested that it actually works):

```
function dashboard_user_login(&$edit, $account) {
  // Don't redirect if a destination is already set.
  if (!empty($_GET['destination'])) {
    return;
  }
  // Don't redirect for password reset.
  if (isset($_POST['form_id']) && $_POST['form_id'] == 'user_pass_reset') {
    return;
  }
  // Check dashboard redirect setting.
  if (config_get('dashboard.settings', 'dashboard_login_redirect') != 1) {
    return;
  }
  // Check for permission to use dashboard.
  if (!user_access('access dashboard', $account)) {
    return;
  }
  // Check for permission to access dashboard layout.
  $dashboard = layout_load('dashboard');
  if ($dashboard && ($dashboard->disabled || !$dashboard->checkAccess())) {
    return;
  }
  $_GET['destination'] = 'admin/dashboard';
}
```

**Comment by quicksketch** [comment]
New sandbox.

File: core/modules/dashboard/dashboard.module [hunk 6]
```diff
@@ -12,20 +12,34 @@ function dashboard_menu() {
...
   }
 
   return $items;
 }
 
+/**
+ * Implements hook_admin_bar_output_alter().
+ */
+function dashboard_admin_bar_output_alter(&$content) {
+  if (isset($content['menu']['menu']['admin/dashboard']['admin/dashboard/default'])) {
```
Comment (klonos) [inline core/modules/dashboard/dashboard.module:38]: It is not clear to me when reading this why it is here for. Can we please add a comment?

File: core/modules/dashboard/dashboard.module (hunk 5 above)
Comment (jenlampton) [reply core/modules/dashboard/dashboard.module:147]: Fixed. (will push a new PR shortly)

File: core/modules/dashboard/dashboard.module (hunk 6 above)
Comment (jenlampton) [reply core/modules/dashboard/dashboard.module:38]: Fixed (will push a new PR in a minute)

**Comment by jenlampton** [comment]
Replaced by https://github.com/backdrop/backdrop/pull/2524

**Pull request closed without merging**